### ISSUE TITLE
log-courier+zmq performance when one of the server gets stuck

### ISSUE THREAD

**Issue by sysmonk** [driskell/log-courier#66]
Hi,

I recently started using ZMQ on a few nodes, but happen to hit some issues. I described one of them in https://github.com/driskell/log-courier/issues/65.
When logstash gets stuck, my log-courier nodes seem to get stuck too. Here are some logs:

```
Nov  4 11:01:21 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:01:21 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:01:21 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:01:21 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:01:21 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:01:21 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:02:21 logger1 log-courier[22765]: Transport error, will try again: Server did not respond within network timeout
Nov  4 11:02:41 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:02:41 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:02:41 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:02:41 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:02:41 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:02:41 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:02:41 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:02:41 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:02:41 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:02:41 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:02:41 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:02:41 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:02:41 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:02:41 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:02:41 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:02:41 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:02:41 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:02:41 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:02:41 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:02:41 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:02:41 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:02:41 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:03:41 logger1 log-courier[22765]: Transport error, will try again: Server did not respond within network timeout
Nov  4 11:04:01 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:04:01 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:04:01 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:04:01 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:04:01 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:04:01 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:04:01 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:04:01 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:04:01 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:04:01 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:04:01 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:04:01 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:04:01 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:04:01 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:04:01 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:04:01 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:04:02 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:04:02 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:04:02 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:04:02 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:04:02 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:04:02 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:05:02 logger1 log-courier[22765]: Transport error, will try again: Server did not respond within network timeout
Nov  4 11:05:22 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:05:22 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:05:22 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:05:22 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:05:22 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:05:22 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:05:22 logger1 log-courier[22765]: Registrar received offsets for 128 log entries
Nov  4 11:05:22 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
Nov  4 11:05:22 logger1 log-courier[22765]: Spooler flushing 128 events due to spool size reached
```

This happens until i restart the stale logstash servers.

log-courier config file:

```
{
        "general": {
                "spool timeout": 5,
                "spool size": 128,
                "log level" : "debug",
                "log syslog" : true,
                "admin enabled": true,
                "admin listen address": "unix:/var/run/log-courier-admin.socket"
        },
        "network": {
                "transport": "plainzmq",
                "servers": [ 
                                        "logstash1.dc-1", "logstash2.dc-1", "logstash3.dc-1", "logstash4.dc-1", 
                                        "logstash1.dc-2", "logstash2.dc-2", "logstash3.dc-2", "logstash4.dc-2",
                                        "logstash1.dc-3", "logstash2.dc-3", "logstash3.dc-3", "logstash4.dc-3"
                                        ],
                "timeout": 10,
                "reconnect": 20,
                "max pending payloads": 5
        },
        "includes": [
                "/etc/log-courier.d/*.conf"
        ]
}
```

Log-courier versoin: ```Log Courier version v1.1.1.gcc82af2```

So, looks like when one of the hosts is down/gets stuck, it still tries to send to it. I thought that if i have 5 payloads, it would send them randomly to servers, rather than all 5 to one server, then the batch of next 5 - to other server and etc. I understand that this is probably more of an ZMQ issue, but maybe there's something that can be done from log-courier side?

**Comment by driskell** [comment]
I've got a redesign of the ZMQ layer penned out which will solve all this. Side effect is TCP/TLS should then connect multiple logstash too :-)

**Comment by driskell** [comment]
@sysmonk Hey. I've decided that I'm going to remove ZeroMQ support from the next major version of Log Courier (2.0). I'm refactoring the entire publisher/transport layer (which is the fix I mentioned before for this) so it can connect to multiple endpoints simultaneously. This would work over TCP/TLS or any other transport (and I'm looking at Kafka atm), and allow Log Courier to effectively manage all queues, meaning things sent to a dead server can be pulled off its queue and resent elsewhere immediately.

This reduces the benefit of keeping ZeroMQ significantly, and even if I kept ZeroMQ there are other queueing related issues that would still be unresolved as ultimately it will still have an underlying queue we can't get at. There's also the fact gozmq is now unmaintained (the library I used) and the only reasonable replacement I can find (pebbe/zmq) does not allow a single library for both ZMQ3 and ZMQ4, meaning I'd have to duplicate code for different ZeroMQ builds, or support my own library. Furthermore, adoption of the ZMQ4 libraries is still slow in major Linux distributions.

If you do have any questions, let me know. Sorry this issue was standing so long too. Hopefully switch from ZeroMQ to TCP should be fine (or from CurveZMQ to TLS.) Let me know if you have concerns here though.
